In [1]:
!pip install ipython-sql

%load_ext sql
        
%sql postgresql://postgres:{pw}@localhost:5432/northwind

In [2]:
%%sql 
SELECT table_name AS name,
       table_type AS type
  FROM information_schema.tables
 WHERE table_schema = 'public' AND table_type IN ('BASE TABLE', 'VIEW');

 * postgresql://postgres:***@localhost:5432/northwind
17 rows affected.


[('territories', 'BASE TABLE'),
 ('order_details', 'BASE TABLE'),
 ('employee_territories', 'BASE TABLE'),
 ('us_states', 'BASE TABLE'),
 ('customers', 'BASE TABLE'),
 ('orders', 'BASE TABLE'),
 ('employees', 'BASE TABLE'),
 ('shippers', 'BASE TABLE'),
 ('products', 'BASE TABLE'),
 ('categories', 'BASE TABLE'),
 ('suppliers', 'BASE TABLE'),
 ('region', 'BASE TABLE'),
 ('customer_demographics', 'BASE TABLE'),
 ('customer_customer_demo', 'BASE TABLE'),
 ('orders_with_customer', 'VIEW'),
 ('detailed_order', 'VIEW'),
 ('in_charge_of', 'VIEW')]

In [19]:
%%sql -- creating a view

CREATE  OR REPLACE VIEW orders_with_customer AS 
SELECT ord.* , cust.customer_id AS customer_table_cust_id,cust.company_name AS comapny_name,cust.country AS country
  FROM orders AS ord 
  JOIN customers AS cust
    ON ord.customer_id = cust.customer_id;


 * postgresql://postgres:***@localhost:5432/northwind
Done.


[]

In [3]:
%%sql
SELECT * 
  FROM orders_with_customer
 LIMIT 5;

 * postgresql://postgres:***@localhost:5432/northwind
5 rows affected.


[(10248, 'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, 'Vins et alcools Chevalier', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France', 'VINET', 'Vins et alcools Chevalier', 'France'),
 (10249, 'TOMSP', 6, datetime.date(1996, 7, 5), datetime.date(1996, 8, 16), datetime.date(1996, 7, 10), 1, 11.61, 'Toms Spezialitäten', 'Luisenstr. 48', 'Münster', None, '44087', 'Germany', 'TOMSP', 'Toms Spezialitäten', 'Germany'),
 (10250, 'HANAR', 4, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 12), 2, 65.83, 'Hanari Carnes', 'Rua do Paço, 67', 'Rio de Janeiro', 'RJ', '05454-876', 'Brazil', 'HANAR', 'Hanari Carnes', 'Brazil'),
 (10251, 'VICTE', 3, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 15), 1, 41.34, 'Victuailles en stock', '2, rue du Commerce', 'Lyon', None, '69004', 'France', 'VICTE', 'Victuailles en stock', 'France'),
 (10252, 'SUPRD', 4, datetime.date(1996, 7, 9), datetime.date(1996, 8, 6), datetime.date(1996, 7, 11), 2, 51.3, 'Suprêmes délices', 'Boulevard Tirou, 255', 'Charleroi', None, 'B-6000', 'Belgium', 'SUPRD', 'Suprêmes délices', 'Belgium')]

In [33]:
%%sql -- created a detailed order 
DROP VIEW detailed_order;
CREATE VIEW detailed_order AS 
SELECT odr.*,pdr.product_name,pdr.product_id,pdr.category_id, odr_deets.quantity,odr_deets.unit_price,odr_deets.discount,CASE
                                                                       WHEN odr_deets.discount > 0 THEN 
                                                                                                 ROUND((1.0 -odr_deets.discount):: numeric 
                                                                                                    *odr_deets.quantity 
                                                                                                       * odr_deets.unit_price :: numeric,2)
                                                                       ELSE ROUND(odr_deets.quantity * odr_deets.unit_price::numeric,2) 
                                                                       END AS sales
  FROM orders AS odr
  JOIN order_details as odr_deets
    ON odr.order_id = odr_deets.order_id
  JOIN products AS pdr
    ON odr_deets.product_id = pdr.product_id
    
;

 * postgresql://postgres:***@localhost:5432/northwind
Done.
Done.


[]

In [34]:
%%sql
SELECT * 
  FROM detailed_order
 LIMIT 15;

 * postgresql://postgres:***@localhost:5432/northwind
15 rows affected.


[(10248, 'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, 'Vins et alcools Chevalier', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France', 'Queso Cabrales', 11, 4, 12, 14.0, 0.0, Decimal('168.00')),
 (10248, 'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, 'Vins et alcools Chevalier', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France', 'Singaporean Hokkien Fried Mee', 42, 5, 10, 9.8, 0.0, Decimal('98.00')),
 (10248, 'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, 'Vins et alcools Chevalier', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France', 'Mozzarella di Giovanni', 72, 4, 5, 34.8, 0.0, Decimal('174.00')),
 (10249, 'TOMSP', 6, datetime.date(1996, 7, 5), datetime.date(1996, 8, 16), datetime.date(1996, 7, 10), 1, 11.61, 'Toms Spezialitäten', 'Luisenstr. 48', 'Münster', None, '44087', 'Germany', 'Tofu', 14, 7, 9, 18.6, 0.0, Decimal('167.40')),
 (10249, 'TOMSP', 6, datetime.date(1996, 7, 5), datetime.date(1996, 8, 16), datetime.date(1996, 7, 10), 1, 11.61, 'Toms Spezialitäten', 'Luisenstr. 48', 'Münster', None, '44087', 'Germany', 'Manjimup Dried Apples', 51, 7, 40, 42.4, 0.0, Decimal('1696.00')),
 (10250, 'HANAR', 4, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 12), 2, 65.83, 'Hanari Carnes', 'Rua do Paço, 67', 'Rio de Janeiro', 'RJ', '05454-876', 'Brazil', "Jack's New England Clam Chowder", 41, 8, 10, 7.7, 0.0, Decimal('77.00')),
 (10250, 'HANAR', 4, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 12), 2, 65.83, 'Hanari Carnes', 'Rua do Paço, 67', 'Rio de Janeiro', 'RJ', '05454-876', 'Brazil', 'Manjimup Dried Apples', 51, 7, 35, 42.4, 0.15, Decimal('1261.40')),
 (10250, 'HANAR', 4, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 12), 2, 65.83, 'Hanari Carnes', 'Rua do Paço, 67', 'Rio de Janeiro', 'RJ', '05454-876', 'Brazil', 'Louisiana Fiery Hot Pepper Sauce', 65, 2, 15, 16.8, 0.15, Decimal('214.20')),
 (10251, 'VICTE', 3, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 15), 1, 41.34, 'Victuailles en stock', '2, rue du Commerce', 'Lyon', None, '69004', 'France', "Gustaf's Knäckebröd", 22, 5, 6, 16.8, 0.05, Decimal('95.76')),
 (10251, 'VICTE', 3, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 15), 1, 41.34, 'Victuailles en stock', '2, rue du Commerce', 'Lyon', None, '69004', 'France', 'Ravioli Angelo', 57, 5, 15, 15.6, 0.05, Decimal('222.30')),
 (10251, 'VICTE', 3, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 15), 1, 41.34, 'Victuailles en stock', '2, rue du Commerce', 'Lyon', None, '69004', 'France', 'Louisiana Fiery Hot Pepper Sauce', 65, 2, 20, 16.8, 0.0, Decimal('336.00')),
 (10252, 'SUPRD', 4, datetime.date(1996, 7, 9), datetime.date(1996, 8, 6), datetime.date(1996, 7, 11), 2, 51.3, 'Suprêmes délices', 'Boulevard Tirou, 255', 'Charleroi', None, 'B-6000', 'Belgium', "Sir Rodney's Marmalade", 20, 3, 40, 64.8, 0.05, Decimal('2462.40')),
 (10252, 'SUPRD', 4, datetime.date(1996, 7, 9), datetime.date(1996, 8, 6), datetime.date(1996, 7, 11), 2, 51.3, 'Suprêmes délices', 'Boulevard Tirou, 255', 'Charleroi', None, 'B-6000', 'Belgium', 'Geitost', 33, 4, 25, 2.0, 0.05, Decimal('47.50')),
 (10252, 'SUPRD', 4, datetime.date(1996, 7, 9), datetime.date(1996, 8, 6), datetime.date(1996, 7, 11), 2, 51.3, 'Suprêmes délices', 'Boulevard Tirou, 255', 'Charleroi', None, 'B-6000', 'Belgium', 'Camembert Pierrot', 60, 4, 40, 27.2, 0.0, Decimal('1088.00')),
 (10253, 'HANAR', 3, datetime.date(1996, 7, 10), datetime.date(1996, 7, 24), datetime.date(1996, 7, 16), 2, 58.17, 'Hanari Carnes', 'Rua do Paço, 67', 'Rio de Janeiro', 'RJ', '05454-876', 'Brazil', 'Gorgonzola Telino', 31, 4, 20, 10.0, 0.0, Decimal('200.00'))]

In [59]:
%%sql
CREATE VIEW in_charge_of AS 
SELECT c.*, e.employee_id, e.last_name || ' '|| e.first_name AS full_name
  FROM customers AS c
  JOIN orders AS o
    ON c.customer_id = o.customer_id
  JOIN employees AS e
    ON o.employee_id = e.employee_id
;


 * postgresql://postgres:***@localhost:5432/northwind
Done.


[]

In [3]:
%%sql
SELECT * 
  FROM in_charge_of
 LIMIT 5;

 * postgresql://postgres:***@localhost:5432/northwind
5 rows affected.


[(10248, 'VINET', 5, datetime.date(1996, 7, 4), datetime.date(1996, 8, 1), datetime.date(1996, 7, 16), 3, 32.38, 'Vins et alcools Chevalier', "59 rue de l'Abbaye", 'Reims', None, '51100', 'France', 'Buchanan Steven'),
 (10249, 'TOMSP', 6, datetime.date(1996, 7, 5), datetime.date(1996, 8, 16), datetime.date(1996, 7, 10), 1, 11.61, 'Toms Spezialitäten', 'Luisenstr. 48', 'Münster', None, '44087', 'Germany', 'Suyama Michael'),
 (10250, 'HANAR', 4, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 12), 2, 65.83, 'Hanari Carnes', 'Rua do Paço, 67', 'Rio de Janeiro', 'RJ', '05454-876', 'Brazil', 'Peacock Margaret'),
 (10251, 'VICTE', 3, datetime.date(1996, 7, 8), datetime.date(1996, 8, 5), datetime.date(1996, 7, 15), 1, 41.34, 'Victuailles en stock', '2, rue du Commerce', 'Lyon', None, '69004', 'France', 'Leverling Janet'),
 (10252, 'SUPRD', 4, datetime.date(1996, 7, 9), datetime.date(1996, 8, 6), datetime.date(1996, 7, 11), 2, 51.3, 'Suprêmes délices', 'Boulevard Tirou, 255', 'Charleroi', None, 'B-6000', 'Belgium', 'Peacock Margaret')]

## Ranking of employees by Sale 

In [50]:
%%sql
WITH -- iii)calculate the ranking of employees by sale 
employees_ranked AS 
(
SELECT sales_tbl.employee_id , sales_tbl.full_name, SUM(sales_tbl.sales) AS employee_sales
       
 FROM (SELECT deo.* ,icf.full_name 
         FROM detailed_order AS deo
        INNER JOIN in_charge_of AS icf
           ON deo.order_id = icf.order_id
        ) AS sales_tbl
 GROUP BY sales_tbl.employee_id, sales_tbl.full_name

)
SELECT * ,RANK() OVER(ORDER BY employee_sales DESC) AS ranks
  FROM employees_ranked
 ;


 * postgresql://postgres:***@localhost:5432/northwind
9 rows affected.


[(4, 'Peacock Margaret', Decimal('232890.85'), 1),
 (3, 'Leverling Janet', Decimal('202812.84'), 2),
 (1, 'Davolio Nancy', Decimal('192107.57'), 3),
 (2, 'Fuller Andrew', Decimal('166537.75'), 4),
 (8, 'Callahan Laura', Decimal('126862.28'), 5),
 (7, 'King Robert', Decimal('124568.23'), 6),
 (9, 'Dodsworth Anne', Decimal('77308.04'), 7),
 (6, 'Suyama Michael', Decimal('73913.13'), 8),
 (5, 'Buchanan Steven', Decimal('68792.25'), 9)]

Peacock Margret is the best salesperson 

## Running Total of Monthly Sales

In [58]:
%%sql
--Running Total Of Monthly Sales part 4

SELECT DATE_TRUNC('month', order_date) :: DATE AS month
        ,SUM(sales) AS monthly_sales
        ,SUM(SUM(sales))OVER(ORDER BY DATE_TRUNC('month', order_date) ) AS running_monthly_sales
  FROM detailed_order as dor
  GROUP BY DATE_TRUNC('month', order_date)
  ;

 * postgresql://postgres:***@localhost:5432/northwind
23 rows affected.


[(datetime.date(1996, 7, 1), Decimal('27861.89'), Decimal('27861.89')),
 (datetime.date(1996, 8, 1), Decimal('25485.28'), Decimal('53347.17')),
 (datetime.date(1996, 9, 1), Decimal('26381.40'), Decimal('79728.57')),
 (datetime.date(1996, 10, 1), Decimal('37515.72'), Decimal('117244.29')),
 (datetime.date(1996, 11, 1), Decimal('45600.04'), Decimal('162844.33')),
 (datetime.date(1996, 12, 1), Decimal('45239.63'), Decimal('208083.96')),
 (datetime.date(1997, 1, 1), Decimal('61258.07'), Decimal('269342.03')),
 (datetime.date(1997, 2, 1), Decimal('38483.63'), Decimal('307825.66')),
 (datetime.date(1997, 3, 1), Decimal('38547.21'), Decimal('346372.87')),
 (datetime.date(1997, 4, 1), Decimal('53032.95'), Decimal('399405.82')),
 (datetime.date(1997, 5, 1), Decimal('53781.28'), Decimal('453187.10')),
 (datetime.date(1997, 6, 1), Decimal('36362.80'), Decimal('489549.90')),
 (datetime.date(1997, 7, 1), Decimal('51020.83'), Decimal('540570.73')),
 (datetime.date(1997, 8, 1), Decimal('47287.66'), Decimal('587858.39')),
 (datetime.date(1997, 9, 1), Decimal('55629.24'), Decimal('643487.63')),
 (datetime.date(1997, 10, 1), Decimal('66749.23'), Decimal('710236.86')),
 (datetime.date(1997, 11, 1), Decimal('43533.79'), Decimal('753770.65')),
 (datetime.date(1997, 12, 1), Decimal('71398.42'), Decimal('825169.07')),
 (datetime.date(1998, 1, 1), Decimal('94222.12'), Decimal('919391.19')),
 (datetime.date(1998, 2, 1), Decimal('99415.29'), Decimal('1018806.48')),
 (datetime.date(1998, 3, 1), Decimal('104854.14'), Decimal('1123660.62')),
 (datetime.date(1998, 4, 1), Decimal('123798.70'), Decimal('1247459.32')),
 (datetime.date(1998, 5, 1), Decimal('18333.62'), Decimal('1265792.94'))]

Total Lifetime Sales is **$1,265,792.94**

# Month over Month Sales Growth

In [62]:
%%sql 
--Month-Over_Month Sales Growth
WITH 
total_monthly_sales AS
(
  SELECT DATE_TRUNC('month', order_date) :: DATE AS month
        ,SUM(sales) AS monthly_sales
    FROM detailed_order
   GROUP BY DATE_TRUNC('month', order_date)
)
,
tms_with_lastmonth_sales AS 
(
  SELECT tms.*, LAG(tms.monthly_sales,1)OVER(ORDER BY tms.month) AS previous_month
    FROM total_monthly_sales AS tms
    


)


SELECT * , ROUND((monthly_sales - previous_month) / previous_month * 100,3) AS pct_growth_rate
  FROM tms_with_lastmonth_sales;

 * postgresql://postgres:***@localhost:5432/northwind
23 rows affected.


[(datetime.date(1996, 7, 1), Decimal('27861.89'), None, None),
 (datetime.date(1996, 8, 1), Decimal('25485.28'), Decimal('27861.89'), Decimal('-8.530')),
 (datetime.date(1996, 9, 1), Decimal('26381.40'), Decimal('25485.28'), Decimal('3.516')),
 (datetime.date(1996, 10, 1), Decimal('37515.72'), Decimal('26381.40'), Decimal('42.205')),
 (datetime.date(1996, 11, 1), Decimal('45600.04'), Decimal('37515.72'), Decimal('21.549')),
 (datetime.date(1996, 12, 1), Decimal('45239.63'), Decimal('45600.04'), Decimal('-0.790')),
 (datetime.date(1997, 1, 1), Decimal('61258.07'), Decimal('45239.63'), Decimal('35.408')),
 (datetime.date(1997, 2, 1), Decimal('38483.63'), Decimal('61258.07'), Decimal('-37.178')),
 (datetime.date(1997, 3, 1), Decimal('38547.21'), Decimal('38483.63'), Decimal('0.165')),
 (datetime.date(1997, 4, 1), Decimal('53032.95'), Decimal('38547.21'), Decimal('37.579')),
 (datetime.date(1997, 5, 1), Decimal('53781.28'), Decimal('53032.95'), Decimal('1.411')),
 (datetime.date(1997, 6, 1), Decimal('36362.80'), Decimal('53781.28'), Decimal('-32.388')),
 (datetime.date(1997, 7, 1), Decimal('51020.83'), Decimal('36362.80'), Decimal('40.311')),
 (datetime.date(1997, 8, 1), Decimal('47287.66'), Decimal('51020.83'), Decimal('-7.317')),
 (datetime.date(1997, 9, 1), Decimal('55629.24'), Decimal('47287.66'), Decimal('17.640')),
 (datetime.date(1997, 10, 1), Decimal('66749.23'), Decimal('55629.24'), Decimal('19.989')),
 (datetime.date(1997, 11, 1), Decimal('43533.79'), Decimal('66749.23'), Decimal('-34.780')),
 (datetime.date(1997, 12, 1), Decimal('71398.42'), Decimal('43533.79'), Decimal('64.007')),
 (datetime.date(1998, 1, 1), Decimal('94222.12'), Decimal('71398.42'), Decimal('31.967')),
 (datetime.date(1998, 2, 1), Decimal('99415.29'), Decimal('94222.12'), Decimal('5.512')),
 (datetime.date(1998, 3, 1), Decimal('104854.14'), Decimal('99415.29'), Decimal('5.471')),
 (datetime.date(1998, 4, 1), Decimal('123798.70'), Decimal('104854.14'), Decimal('18.068')),
 (datetime.date(1998, 5, 1), Decimal('18333.62'), Decimal('123798.70'), Decimal('-85.191'))]

Biggest growth month over month is from **November 1997 to December 1997** at about **64%** <br>

Largest decrease in growth month over month is from **April 1998 to May 1998** at **-85.19%**

### Identifying High Value Customers

In [63]:
%%sql
-- part(vi)
WITH sum_of_sales_by_order_id
AS
(
SELECT customer_id, order_id,SUM(sales) AS sum_sales_by_ord_id
  FROM detailed_order
    GROUP BY order_id,customer_id
)
,
/*above_avg_counter
AS 
(
 SELECT cutsomer_id, COUNT(*)
   FROM custom*/
customer_aov_tbl 
AS 
(
SELECT *, AVG(sum_sales_by_ord_id) OVER(ORDER BY customer_id
                                        ROWS  BETWEEN  UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING )AS average,
                                     CASE
                                     WHEN sum_sales_by_ord_id > AVG(sum_sales_by_ord_id) OVER() THEN 'Above Average'
                                     ELSE 'Below Average'
                                     END AS average_flag
                                    
                               
                                 
  FROM sum_of_sales_by_order_id
    
)
    
 SELECT * , COUNT(CASE WHEN average_flag = 'Above Average' THEN 1 ELSE NULL END) OVER(PARTITION BY customer_id) AS ord_abv_avg
   FROM customer_aov_tbl
    ORDER BY ord_abv_avg DESC, customer_id
    ;

 * postgresql://postgres:***@localhost:5432/northwind
830 rows affected.


[('ERNSH', 11072, Decimal('5218.00'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10633, Decimal('5510.59'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 11017, Decimal('6750.00'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10836, Decimal('4705.50'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10764, Decimal('2286.00'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10990, Decimal('4288.85'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10854, Decimal('2966.50'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10402, Decimal('2713.50'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10773, Decimal('2030.40'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10390, Decimal('2090.88'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 11008, Decimal('4680.90'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10771, Decimal('344.00'), Decimal('1525.0517349397590361'), 'Below Average', 26),
 ('ERNSH', 10351, Decimal('5398.72'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10776, Decimal('6635.27'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10430, Decimal('4899.20'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10667, Decimal('1536.80'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10403, Decimal('855.01'), Decimal('1525.0517349397590361'), 'Below Average', 26),
 ('ERNSH', 10595, Decimal('4725.00'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10263, Decimal('1873.80'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10442, Decimal('1792.00'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10514, Decimal('8623.45'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10382, Decimal('2900.00'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10698, Decimal('3436.45'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10258, Decimal('1614.88'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10979, Decimal('4813.50'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10368, Decimal('1689.78'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10895, Decimal('6379.40'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10795, Decimal('2158.00'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('ERNSH', 10571, Decimal('550.59'), Decimal('1525.0517349397590361'), 'Below Average', 26),
 ('ERNSH', 10968, Decimal('1408.00'), Decimal('1525.0517349397590361'), 'Below Average', 26),
 ('SAVEA', 10657, Decimal('4371.60'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('SAVEA', 10612, Decimal('6375.00'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('SAVEA', 10714, Decimal('2205.75'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('SAVEA', 10748, Decimal('2196.00'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('SAVEA', 10722, Decimal('1570.00'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('SAVEA', 10983, Decimal('720.90'), Decimal('1525.0517349397590361'), 'Below Average', 26),
 ('SAVEA', 10398, Decimal('2505.60'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('SAVEA', 10882, Decimal('892.64'), Decimal('1525.0517349397590361'), 'Below Average', 26),
 ('SAVEA', 10678, Decimal('5256.50'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('SAVEA', 10847, Decimal('4931.92'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('SAVEA', 10452, Decimal('2018.50'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('SAVEA', 10894, Decimal('2753.10'), Decimal('1525.0517349397590361'), 'Above Average', 26),
 ('SAVEA', 10757, Decimal('3082.00'), Decimal('1525.05173

Testing


### Percentage of sales within each category

In [49]:
%%sql 
-- part vii percentage of sales within each category
WITH 
cat_sales_tbl AS
(

SELECT c.category_id, c.category_name, SUM(dto.SALES) AS sales_by_cat
  FROM detailed_order AS dto
  JOIN categories AS c
    ON dto.category_id = c.category_id
 GROUP BY c.category_id , c.category_name
)
SELECT * , ROUND((sales_by_cat 
           /SUM(sales_by_cat)
            OVER(ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING )) * 100,2) AS pct_of_sales
  FROM cat_sales_tbl

 * postgresql://postgres:***@localhost:5432/northwind
8 rows affected.


[(8, 'Seafood', Decimal('131261.73'), Decimal('10.37')),
 (7, 'Produce', Decimal('99984.57'), Decimal('7.90')),
 (1, 'Beverages', Decimal('267868.16'), Decimal('21.16')),
 (5, 'Grains/Cereals', Decimal('95744.59'), Decimal('7.56')),
 (2, 'Condiments', Decimal('106047.07'), Decimal('8.38')),
 (4, 'Dairy Products', Decimal('234507.26'), Decimal('18.53')),
 (6, 'Meat/Poultry', Decimal('163022.37'), Decimal('12.88')),
 (3, 'Confections', Decimal('167357.19'), Decimal('13.22'))]

Beverages make up most of the sales at 21.16% followed by dairy products at 18.53% and then meat/poultry at 12.88%


### Find the top 3 products sold in the  country

In [61]:
%%sql
WITH
tt_sales_prd AS --sales by product id
(
SELECT product_id
      , product_name,category_id
      , SUM(SALES) AS total_sales
      , ROW_NUMBER() OVER(PARTITION BY category_id ORDER BY SUM(SALES) DESC) AS row_num
  FROM detailed_order
 GROUP BY product_id,product_name, category_id
)

SELECT * 
  FROM tt_sales_prd
 WHERE row_num <= 3
 
    

 * postgresql://postgres:***@localhost:5432/northwind
24 rows affected.


[(38, 'Côte de Blaye', 1, Decimal('141396.73'), 1),
 (43, 'Ipoh Coffee', 1, Decimal('23526.70'), 2),
 (2, 'Chang', 1, Decimal('16355.96'), 3),
 (63, 'Vegie-spread', 2, Decimal('16701.10'), 1),
 (61, "Sirop d'érable", 2, Decimal('14352.60'), 2),
 (65, 'Louisiana Fiery Hot Pepper Sauce', 2, Decimal('13869.89'), 3),
 (62, 'Tarte au sucre', 3, Decimal('47234.98'), 1),
 (20, "Sir Rodney's Marmalade", 3, Decimal('22563.36'), 2),
 (26, 'Gumbär Gummibärchen', 3, Decimal('19849.13'), 3),
 (59, 'Raclette Courdavault', 4, Decimal('71155.70'), 1),
 (60, 'Camembert Pierrot', 4, Decimal('46825.48'), 2),
 (72, 'Mozzarella di Giovanni', 4, Decimal('24900.13'), 3),
 (56, 'Gnocchi di nonna Alice', 5, Decimal('42593.06'), 1),
 (64, 'Wimmers gute Semmelknödel', 5, Decimal('21957.97'), 2),
 (42, 'Singaporean Hokkien Fried Mee', 5, Decimal('8575.00'), 3),
 (29, 'Thüringer Rostbratwurst', 6, Decimal('80368.69'), 1),
 (17, 'Alice Mutton', 6, Decimal('32698.38'), 2),
 (53, 'Perth Pasties', 6, Decimal('20574.17'), 3),
 (51, 'Manjimup Dried Apples', 7, Decimal('41819.65'), 1),
 (28, 'Rössle Sauerkraut', 7, Decimal('25696.64'), 2),
 (7, "Uncle Bob's Organic Dried Pears", 7, Decimal('22044.30'), 3),
 (18, 'Carnarvon Tigers', 8, Decimal('29171.88'), 1),
 (10, 'Ikura', 8, Decimal('20867.34'), 2),
 (40, 'Boston Crab Meat', 8, Decimal('17910.63'), 3)]

This query *shows* the **top 3 products** in sales by category